File: frontend/src/lib/api/checkpoints.ts
/**
 * Typed client for the checkpoints REST surface (G6;
 * ``src/bearings/web/routes/checkpoints.py``).
 *
 * Mirrors :class:`bearings.web.models.checkpoints.CheckpointOut` /
 * :class:`bearings.web.models.checkpoints.CheckpointIn` /
 * :class:`bearings.web.models.checkpoints.CheckpointForkResult` field
 * for field. The conversation pane fetches the per-session list on
 * mount (and re-fetches after a create / delete / fork) to refresh the
 * gutter chips.
 */
import { API_CHECKPOINTS_ENDPOINT, checkpointEndpoint, checkpointForkEndpoint } from "../config";
import { deleteResource, getJson, postJson, type RequestOptions } from "./client";

/** Wire shape — one-to-one with :class:`bearings.web.models.checkpoints.CheckpointOut`. */
export interface CheckpointOut {
  id: string;
  session_id: string;
  message_id: string;
  label: string;
  created_at: string;
}

/**
 * Wire envelope for ``POST /api/checkpoints/{id}/fork``. Consumed via
 * the ``forkCheckpoint`` return type — not exported because no caller
 * imports the type independently of the function (knip gate).
 */
interface CheckpointForkResult {
  new_session_id: string;
  source_session_id: string;
  checkpoint_id: string;
  message_count: number;
}

interface CreateCheckpointParams {
  sessionId: string;
  messageId: string;
  /** Omit to let the server synthesise a default label. */
  label?: string;
}

/**
 * Create a checkpoint at ``messageId``. Returns the new row.
 *
 * The ``label`` is optional — when omitted the route synthesises one
 * from :data:`bearings.config.constants.DEFAULT_CHECKPOINT_LABEL_TEMPLATE`.
 */
export async function createCheckpoint(
  params: CreateCheckpointParams,
  options: RequestOptions = {},
): Promise<CheckpointOut> {
  const body: Record<string, string> = {
    session_id: params.sessionId,
    message_id: params.messageId,
  };
  if (params.label !== undefined && params.label.length > 0) {
    body.label = params.label;
  }
  return await postJson<CheckpointOut>(API_CHECKPOINTS_ENDPOINT, body, options);
}

/**
 * List every checkpoint for ``sessionId``, newest-first. Returns ``[]``
 * for an unknown / empty session — the gutter renders nothing in
 * either case.
 */
export async function listCheckpoints(
  sessionId: string,
  options: RequestOptions = {},
): Promise<CheckpointOut[]> {
  return await getJson<CheckpointOut[]>(API_CHECKPOINTS_ENDPOINT, {
    ...options,
    query: [["session_id", sessionId]],
  });
}

/**
 * Delete one checkpoint. 204 No Content on success; 404 from the server
 * surfaces as :class:`ApiError`.
 */
export async function deleteCheckpoint(
  checkpointId: string,
  options: RequestOptions = {},
): Promise<void> {
  return await deleteResource<void>(checkpointEndpoint(checkpointId), options);
}

/**
 * Fork the source session at ``checkpointId``. Returns the new
 * session id + the count of copied messages so the caller can navigate
 * to the new session.
 */
export async function forkCheckpoint(
  checkpointId: string,
  options: RequestOptions = {},
): Promise<CheckpointForkResult> {
  return await postJson<CheckpointForkResult>(checkpointForkEndpoint(checkpointId), {}, options);
}

File: src/bearings/web/models/checkpoints.py
# mypy: disable-error-code=explicit-any
"""Pydantic wire shapes for ``routes/checkpoints.py`` (G6).

Per ``docs/architecture-v1.md`` §1.1.5 the wire DTOs live alongside
their route module. The shapes mirror :class:`bearings.db.checkpoints.Checkpoint`
plus a thin :class:`CheckpointForkResult` envelope returned by the
``POST /api/checkpoints/{id}/fork`` action.

The ``mypy: disable-error-code=explicit-any`` pragma is the same narrow
carve-out the other ``web/models/*.py`` files make for Pydantic's
metaclass-exposed ``Any`` surface — every public ``BaseModel`` subclass
below has a fully-typed field set.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from bearings.config.constants import CHECKPOINT_LABEL_MAX_LENGTH


class CheckpointIn(BaseModel):
    """Request body for ``POST /api/checkpoints``.

    Validators mirror :class:`bearings.db.checkpoints.Checkpoint.__post_init__`
    so a bad payload fails at the wire boundary with a 422 rather than
    surfacing a 500 from the dataclass downstream. ``model_config``
    forbids extra fields so a typo (e.g. ``messageId`` instead of
    ``message_id``) is rejected loudly.
    """

    model_config = ConfigDict(extra="forbid")

    session_id: str = Field(min_length=1)
    message_id: str = Field(min_length=1)
    label: str | None = Field(
        default=None,
        max_length=CHECKPOINT_LABEL_MAX_LENGTH,
        description=(
            "User-visible chip text. When omitted the route layer "
            "synthesises a default per "
            ":data:`bearings.config.constants.DEFAULT_CHECKPOINT_LABEL_TEMPLATE`."
        ),
    )


class CheckpointOut(BaseModel):
    """Response body for checkpoint endpoints.

    One-to-one with :class:`bearings.db.checkpoints.Checkpoint`.
    """

    model_config = ConfigDict(extra="forbid")

    id: str
    session_id: str
    message_id: str
    label: str
    created_at: str


class CheckpointForkResult(BaseModel):
    """Response body for ``POST /api/checkpoints/{id}/fork``.

    The fork action clones the source session row + replays messages up
    to and including the checkpoint anchor into a new session. The wire
    envelope carries both pieces so the frontend can navigate to the
    new session without a follow-up fetch.
    """

    model_config = ConfigDict(extra="forbid")

    new_session_id: str
    source_session_id: str
    checkpoint_id: str
    message_count: int


__all__ = ["CheckpointForkResult", "CheckpointIn", "CheckpointOut"]
